app/models/tslr_claim.rb

```diff
@@ -205,4 +234,17 @@ def bank_sort_code_must_be_six_digits
     errors.add(:bank_sort_code, "Sort code must contain six digits") \
       if bank_sort_code.present? && normalised_bank_detail(bank_sort_code) !~ /\A\d{6}\z/
   end
+
+  def generate_reference!
+    reference = loop {
+      ref = nanoid
+      break ref unless self.class.exists?(reference: ref)
+    }
+
+    update_attribute(:reference, reference)
+  end
+
+  def nanoid
```

Comment:
This method name could be more helpful - it's named after the underlying implementation, not for what it actually does. As discussed, perhaps we can abstract this to a `Reference.generate` class so that it's encapsulated somewhere and we can still stub it in the specs where necessary.